src/nixpacks/builder/docker/dockerfile_generation.rs

```diff
@@ -0,0 +1,253 @@
+use crate::nixpacks::{
+    environment::Environment,
+    images::DEFAULT_BASE_IMAGE,
+    nix,
+    plan::{
+        phase::{Phase, StartPhase},
+        BuildPlan,
+    },
+};
+use anyhow::{Context, Ok, Result};
+use indoc::formatdoc;
+use std::{
+    fs::{self, File},
+    io::Write,
+    path::{Path, PathBuf},
+};
+
+use super::{utils, DockerBuilderOptions};
+
+pub trait DockerfileGenerator {
+    fn generate_dockerfile(
+        &self,
+        options: &DockerBuilderOptions,
+        env: &Environment,
+    ) -> Result<String>;
+    fn write_supporting_files(
+        &self,
+        _options: &DockerBuilderOptions,
+        _env: &Environment,
+        _dest: &str,
+    ) -> Result<()> {
+        Ok(())
+    }
+}
+
+pub static APP_DIR: &str = "/app/";
+
+impl DockerfileGenerator for BuildPlan {
+    fn generate_dockerfile(
+        &self,
+        options: &DockerBuilderOptions,
+        env: &Environment,
+    ) -> Result<String> {
+        let plan = self;
+
+        let variables = plan.variables.clone().unwrap_or_default();
+        let args_string = if !variables.is_empty() {
+            format!(
+                "ARG {}\nENV {}",
+                // Pull the variables in from docker `--build-arg`
+                variables
+                    .iter()
+                    .map(|var| var.0.to_string())
+                    .collect::<Vec<_>>()
+                    .join(" "),
+                // Make the variables available at runtime
+                variables
+                    .iter()
+                    .map(|var| format!("{}=${}", var.0, var.0))
+                    .collect::<Vec<_>>()
+                    .join(" ")
+            )
+        } else {
+            "".to_string()
+        };
+
+        let dockerfile_phases = plan
+            .get_sorted_phases()?
+            .into_iter()
+            .map(|phase| {
+                let phase_dockerfile = phase
+                    .generate_dockerfile(options, env)
+                    .context(format!("Generating Dockerfile for phase {}", phase.name))?;
+
+                match phase.name.as_str() {
+                    // We want to load the variables immediately after the setup phase
+                    "setup" => Ok(format!(
+                        "{phase_dockerfile}\n# load variables\n{args_string}\n"
+                    )),
+                    _ => Ok(phase_dockerfile),
+                }
+            })
+            .collect::<Result<Vec<_>>>();
+        let dockerfile_phases_str = dockerfile_phases?.join("\n");
+
+        let start_phase_str = plan
+            .start_phase
+            .clone()
+            .unwrap_or_default()
+            .generate_dockerfile(options, env)?;
+
+        let base_image = plan.build_image.clone();
+
+        let dockerfile = formatdoc! {"
+            FROM {base_image}
+            WORKDIR {APP_DIR}
+
+            {dockerfile_phases_str}
+
+            {start_phase_str}
+        "};
+
```

Comment:
Could this support docker multi-stage build? (one stage for creating artifact & other stage for running the artifact)